### ISSUE TITLE
Deploy: create a CloudFormation change set and stop there

### ISSUE THREAD

**Issue by ronkorving** [serverless/serverless#10139]
### Is there an existing issue for this?

- [X] I have searched existing issues, it hasn't been reported yet

^ sorry, that is 50% a lie, but I had to tick this button to create the issue.

### Use case description

This has been reported before, in 2017, but was closed for at-the-time probably good reasons. I think it's worth revisiting.

At the time, part of why it was not followed through, seems to be the existence of the [serverless-cloudformation-changesets](https://github.com/trek10inc/serverless-cloudformation-changesets) plugin, which I use myself today. My concern is twofold:

* That plugin is no longer maintained.
* If Serverless introduces breaking changes (ie. the next major version release), I fear that that might be the end of this plugin.

I also think that CloudFormation Change Sets are the kind of feature that deserves to be in core and deserves to gets its maintenance, which is easier to continuously guarantee once it's in core.

### Proposed solution (optional)

Perhaps making CloudFormation Change Sets part of core could be reconsidered?

**Comment by drc-gcoakley** [comment]
They probably restrict it using Control Tower or account permissions. Barring [dscdac](https://github.com/dscdac)'s  changesets as restriction, this functionality is also useful for:
1. auditing the changes that are made at that stack/platform level
   This can be preferable to trying to audit the source code changes which can be much more difficult to follow depending on the method(s) and level of automation used to generate the changes.
2. determining the current state of an environment compared to the current source of truth (latest serverless.yml file)
   We have multiple environments are normally updated in order and/or in sync. But, sometimes that cannot happen because the environment is being used for a non-standard purpose (usually a demo).

**Comment by seanohollaren** [comment]
Any update on the prioritization of this feature, by chance?  We have an identical use case where we like to use change sets that aren't executed automatically for some of our projects which focus largely on defining infrastructure as opposed to code.  It allows us to take the extra step of manually verifying the changes before they're applied.  Since the plugin which used to accomplish this no longer functions, and since serverless is creating change sets now by default, an extra option which would create the change set but stop short of executing it would be really useful for us.

**Comment by laurenshareshian** [comment]
Yes, 👍 to this feature request. More specifically, some sort of conditional [here](https://github.com/serverless/serverless/blob/main/lib/plugins/aws/lib/update-stack.js#L59) would be helpful perhaps in allowing us to have more control over when we choose to execute the change set.

**Comment by astuyve** [comment]
Hi, @dscdac, @mnapoli and others.

I forked `serverless-cloudformation-changesets` and updated it for v3 [here](https://www.npmjs.com/package/serverless-cloudformation-changesets-v3).

If I have more time, I can explore either porting this to v2 or supporting both libraries. I could also propose a PR to upstream this into the framework itself if @medikoo is amenable.

Thanks friends! 💞

**Comment by medikoo** [comment]
@agupta94 this is what Framework since v3 does by default. Deploy is about creating a changeset and applying it.

What's missing is the ability to stop before deployment and let the changeset be deployed from AWS console

**Comment by agupta94** [comment]
@medikoo Right what I am suggesting is NOT through the AWS Console. An option to review and apply through the CLI. At a lot of companies developers don't have direct access to the production environments, and can only deploy through CD tools.

**Comment by agupta94** [comment]
@ronkorving I can think of Terraform (another infrastructure as code language) that provides this sort of workflow with Terraform Plan and Apply functionality.

**Comment by amuroBosetti** [comment]
Hey all! I have some experience setting up CI flows with both Terraform and Serverless, and I can see great value in adding a command that prints everything that is about to change but without actually executing the change just yet. 
- There are certain CI tools (like Jenkins) that can be configured to execute something and wait for an input before continuing the flow. We used to have a deployment workflow with Terraform that would run a plan, wait for ~30 minutes, and if no approval was received, it would cancel the whole execution.
- When we migrated to Github Actions, which doesn't have this 'keep waiting' feature, we simply had two workflows: one that runs a plan, and one that runs an apply. This way, we could trigger the plan, wait for it to be finished, review it, and everything looks ok, only then run the apply. I understand that, with Serverless, this would end up creating two Cloudformation Change Sets, the first of which would never be executed and therefore should be deleted at the end of the plan step. I don't know if that would be a big increment in costs for the user, I can only offer advice regarding the technical aspect.
cc @agupta94, @ronkorving

**Comment by amuroBosetti** [comment]
@medikoo I understand that this is the simplest implementation we can think of, but I'm not sure if it's the most useful, because it would leave out all of the users who are running deploys though a CI tool. I don't know if that's the majority (I would hope so), but I dare say that it's the most common way of deploying for bigger projects.

An alternative that would allow for the workflow described in my last comment would be:
- Introduce a new `serverless plan` CLI command (I'd try to find a different name, so it's not just a copy of Terraform :sweat_smile:).
- It should do the same stuff as a deploy for creating a new change set, ie
  -  On stack creation,  https://github.com/serverless/serverless/blob/4582913214e1c9d6f4324ff6a507dd09a2fd4e1b/lib/plugins/aws/deploy/lib/create-stack.js#L31-L35
  - On stack update: https://github.com/serverless/serverless/blob/4582913214e1c9d6f4324ff6a507dd09a2fd4e1b/lib/plugins/aws/lib/update-stack.js#L100-L104
- We can save o print the ChangeSet description, and after that remove the created ChangeSet just like in https://github.com/serverless/serverless/blob/4582913214e1c9d6f4324ff6a507dd09a2fd4e1b/lib/plugins/aws/lib/update-stack.js#L110-L119

**Comment by medikoo** [comment]
@amuroBosetti so you mean that `serverless plan` should create changeset and remove it?

It technically can work, but it'll be limited, and it'll not fully reflect Terraform plan. Framework in some cases follows with updates made directly via AWS SDK, and this obviously cannot be easily covered under `sls plan`.

**Comment by amuroBosetti** [comment]
@medikoo oh ok, I haven't dived into the code that much, can you point me towards those usages? Also, wouldn't we have the same limitation if we did this change as a CLI param for the `sls deploy` command?

**Comment by medikoo** [comment]
> @medikoo oh ok, I haven't dived into the code that much, can you point me towards those usages? 

API Gateway settings are handled that way:  https://github.com/serverless/serverless/blob/main/lib/plugins/aws/package/compile/events/api-gateway/lib/hack/README.md Still if there's a way to handle that with CF deploy, upgrade is definitely welcome

> Also, wouldn't we have the same limitation if we did this change as a CLI param for the sls deploy command?

We will, still `sls plan` promises full plan, while `--request-confirmation` is simply about injection an interactive confirmation in a middle of a processing.

**Comment by amuroBosetti** [comment]
I see. If we still believe it's something we want to have in Serverless, we can go for the `--request-confirmation` option first, because it would be better than having no plan at all, and leave the actual `sls plan` for the future

**Comment by ronkorving** [comment]
@medikoo Have y'all reached out to the CFN team about any missing capability? You can bring attention to those issues here: https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/issues

**Comment by lautarolecumberry** [comment]
Hi! has there been any updates on this?

**Comment by HarrisonLongartlogic** [comment]
Hey all, Any updates on this? Would be very useful to know the exact changes to the stack before deploying them. 

Cheers.
